src/slurm_plugin/clustermgtd.py

```diff
@@ -904,6 +1030,111 @@ def _terminate_orphaned_instances(self, cluster_instances, ips_used_by_slurm):
+
+    def _find_online_slurm_nodes(self, healthy_nodes):
+        """Find slurm nodes online."""
+        online_nodes = []
+        for node in healthy_nodes:
+            if node.is_online():
+                online_nodes.append(node)
+        return online_nodes
+
+    def _reset_partition_failure_count(self, partition):
+        """Reset bootstrap failure count for partition which has bootstrap failure nodes successfully launched."""
+        log.info(f"Find successfully launched node in partition {partition}, reset partition protected failure count")
+        self._partitions_protected_failure_count_map.pop(partition, None)
+
+    def _handle_bootstrap_failure_nodes(self, bootstrap_failure_nodes):
+        """Increase the partition failure count and add failed nodes to the set of bootstrap failure nodes."""
+        self._increase_partitions_protected_failure_count(bootstrap_failure_nodes)
+        # Update bootstrap_failure_nodes to include all nodes types have been failed during bootstrap
+        self._bootstrap_failure_nodes_types |= get_nodes_type(bootstrap_failure_nodes)
+
+    def _get_online_nodes_types(self, healthy_nodes):
+        """Find nodes types which have been failed during bootstrap now become online."""
+        online_nodes = self._find_online_slurm_nodes(healthy_nodes)
+        online_nodes_types = get_nodes_type(online_nodes)
+
+        return online_nodes_types
+
+    @staticmethod
+    def _get_partitions_to_disable(nodename_to_slurm_nodes_map, bootstrap_failure_partitions, partitions):
+        """Get partitions not running jobs to disable."""
+        # Get partitions have jobs running
+        partitions_have_active_jobs = ClusterManager._filter_partitions_with_active_jobs(
+            nodename_to_slurm_nodes_map, partitions
+        )
+        # Only disable bootstrap failure partitions that not having active jobs running
+        partitions_to_disable = set(bootstrap_failure_partitions) - partitions_have_active_jobs
+        bootstrap_failure_partitions_have_jobs = set(bootstrap_failure_partitions) - partitions_to_disable
+        if bootstrap_failure_partitions_have_jobs:
+            log.info(
+                f"Bootstrap failure partitions {bootstrap_failure_partitions_have_jobs} currently have jobs running, "
+                "not disabling them."
+            )
+        return partitions_to_disable
+
+    def _is_protected_mode_enabled(self):
+        """When protected_failure_count is set to -1, disable protected mode."""
+        if self._config.protected_failure_count == -1:
+            return False
+        return True
+
+    def _handle_protected_mode_process(self, slurm_nodes, partitions, bootstrap_failure_nodes, healthy_nodes):
+        """Handle the process of entering protected mode."""
+        # Handle successfully launched nodes
+        nodename_to_slurm_nodes_map = {node.name: node for node in slurm_nodes}
```

Comment:
You don't need to always initialize this data structure, cause it is only used in the if statement below.
As a general recommendation all these data-structures can contain thousands of nodes based on the potential size of the cluster, which means we should minimize doing all these transformations or copies if possible.